File: orchestrator-api/src/main/java/com/example/gwp/orchestrator/api/dto/PreemptionHistoryResponse.java
package com.example.gwp.orchestrator.api.dto;

import com.example.gwp.orchestrator.domain.JobPriority;
import com.example.gwp.orchestrator.domain.PreemptionHistoryEntry;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

public record PreemptionHistoryResponse(List<Entry> items) {

    public record Entry(
            UUID id,
            UUID victimJobId,
            String victimOwner,
            JobPriority victimPriority,
            int victimGpuCount,
            UUID preemptorJobId,
            String preemptorOwner,
            JobPriority preemptorPriority,
            Instant preemptedAt,
            String reason
    ) {
        public static Entry from(PreemptionHistoryEntry h) {
            return new Entry(
                    h.getId(),
                    h.getVictimJobId(),
                    h.getVictimOwner(),
                    h.getVictimPriority(),
                    h.getVictimGpuCount(),
                    h.getPreemptorJobId(),
                    h.getPreemptorOwner(),
                    h.getPreemptorPriority(),
                    h.getPreemptedAt(),
                    h.getReason()
            );
        }
    }

    public static PreemptionHistoryResponse from(List<PreemptionHistoryEntry> entries) {
        return new PreemptionHistoryResponse(entries.stream().map(Entry::from).toList());
    }
}

File: orchestrator-api/src/main/kotlin/com/example/gwp/orchestrator/api/dto/PreemptionHistoryResponse.kt
package com.example.gwp.orchestrator.api.dto

import com.example.gwp.orchestrator.domain.JobPriority
import com.example.gwp.orchestrator.domain.PreemptionHistoryEntry
import java.time.Instant
import java.util.UUID

@JvmRecord
data class PreemptionHistoryResponse(val items: List<Entry>) {

    @JvmRecord
    data class Entry(
        val id: UUID,
        val victimJobId: UUID,
        val victimOwner: String,
        val victimPriority: JobPriority,
        val victimGpuCount: Int,
        val preemptorJobId: UUID,
        val preemptorOwner: String,
        val preemptorPriority: JobPriority,
        val preemptedAt: Instant,
        val reason: String,
    ) {
        companion object {
            @JvmStatic
            fun from(h: PreemptionHistoryEntry): Entry = Entry(
                h.id,
                h.victimJobId,
                h.victimOwner,
                h.victimPriority,
                h.victimGpuCount,
                h.preemptorJobId,
                h.preemptorOwner,
                h.preemptorPriority,
                h.preemptedAt,
                h.reason,
            )
        }
    }

    companion object {
        @JvmStatic
        fun from(entries: List<PreemptionHistoryEntry>): PreemptionHistoryResponse =
            PreemptionHistoryResponse(entries.map { Entry.from(it) })
    }
}
